# Question 9
We will now consider the Boston housing data set, from the ISLP library.

### (a)  
Based on this data set, provide an estimate for the population mean of \( \text{medv} \). Call this estimate \( \hat{\mu} \).

In [2]:
from ISLP import load_data
import numpy as np

Boston = load_data('Boston')
Boston

,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,lstat,medv
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296,15.3,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242,17.8,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242,17.8,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222,18.7,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222,18.7,5.33,36.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...
501,0.06263,0.0,11.93,0,0.573,6.593,69.1,2.4786,1,273,21.0,9.67,22.4
502,0.04527,0.0,11.93,0,0.573,6.120,76.7,2.2875,1,273,21.0,9.08,20.6
503,0.06076,0.0,11.93,0,0.573,6.976,91.0,2.1675,1,273,21.0,5.64,23.9
504,0.10959,0.0,11.93,0,0.573,6.794,89.3,2.3889,1,273,21.0,6.48,22.0


In [3]:
mu_hat = np.mean(Boston['medv'])
mu_hat

np.float64(22.532806324110677)

### (b)  
Provide an estimate of the standard error of \( \hat{\mu} \). Interpret this result.  
Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations.

In [4]:
se_mu_hat = np.std(Boston['medv'], ddof=1) / np.sqrt(len(Boston['medv']))
se_mu_hat

np.float64(0.40886114749753505)

### (c)  
Now estimate the standard error of \( \hat{\mu} \) using the bootstrap. How does this compare to your answer from (b)?

In [5]:
rng = np.random.default_rng(1)
boot_means = [np.mean(rng.choice(Boston['medv'], size=len(Boston['medv']), replace=True)) for _ in range(1000)]
boot_se = np.std(boot_means, ddof=1)
boot_se

np.float64(0.41554577333267234)

**ANS for question c:**  
The standard error of hat mu from bootstrap (0.4088) is quite the same as we have in question b (0.4155). 

### (d)  
Based on your bootstrap estimate from (c), provide a 95% confidence interval for the mean of \( \text{medv} \). Compare it to the results obtained by using `Boston['medv'].std()` and the two standard error rule (3.9).  
Hint: You can approximate a 95% confidence interval using the formula \([ \hat{\mu} - 2SE(\hat{\mu}), \hat{\mu} + 2SE(\hat{\mu}) ]\).

In [6]:
ci_lower = mu_hat - 2 * boot_se
ci_upper = mu_hat + 2 * boot_se
(ci_lower, ci_upper)

(np.float64(21.701714777445332), np.float64(23.36389787077602))

The 95% confidence interval for this mean of `medv` is [21.7, 23.3]

### (e)  
Based on this data set, provide an estimate, \( \hat{\mu}_{\text{med}} \), for the median value of \( \text{medv} \) in the population.

In [7]:
mu_med_hat = np.median(Boston['medv'])
mu_med_hat

np.float64(21.2)

### (f)  
We now would like to estimate the standard error of \( \hat{\mu}_{\text{med}} \). Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [8]:
boot_medians = [np.median(rng.choice(Boston['medv'], size=len(Boston['medv']), replace=True)) for _ in range(1000)]
boot_se_median = np.std(boot_medians, ddof=1)
boot_se_median

np.float64(0.38278673655925627)

**ANS for question f:**  
We found that standard error of median is higher than standard error of mean.

(g) Based on this data set, provide an estimate for the tenth percentile of \( \text{medv} \) in Boston census tracts. Call this quantity \( \hat{\mu}_{0.1} \).  
(You can use the `np.percentile()` function.)

In [9]:
mu_0_1_hat = np.percentile(Boston['medv'], 10)
mu_0_1_hat

np.float64(12.75)

(h) Use the bootstrap to estimate the standard error of \( \hat{\mu}_{0.1} \). Comment on your findings.

In [10]:
boot_percentiles = [np.percentile(rng.choice(Boston['medv'], size=len(Boston['medv']), replace=True), 10) for _ in range(1000)]
boot_se_percentile = np.std(boot_percentiles, ddof=1)
boot_se_percentile

np.float64(0.5162050247988949)

**ANS for question h:**  
We found that the standard error of hat mu 0.1 (0.5162) is higher than the standard error of hat mu (0.4155).